crates/subspace-farmer/src/single_disk_plot/piece_receiver.rs

```diff
@@ -0,0 +1,236 @@
+
+            match piece_result {
+                Ok(Some(peer_id)) => {
+                    trace!(
+                        %piece_index,
+                        ?key,
+                        ?peer_id,
+                        "get_value returned a piece-by-sector provider"
+                    );
+
+                    if let Ok(piece_provider_id) = PeerId::from_bytes(&peer_id) {
+                        let request_result = dsn_node
+                            .send_generic_request(
+                                piece_provider_id,
+                                PieceByHashRequest {
+                                    key: PieceKey::Sector(PieceIndexHash::from_index(piece_index)),
+                                },
+                            )
+                            .await;
+
+                        match request_result {
+                            Ok(request) => {
+                                if let Some(piece) = request.piece {
+                                    return Some(piece);
+                                }
+                            }
+                            Err(error) => {
+                                error!(%piece_index,?peer_id, ?key, ?error, "Error on piece-by-hash request.");
+                            }
+                        }
+                    } else {
+                        error!(
+                            %piece_index,
+                            ?peer_id,
+                            ?key,
+                            "Cannot convert piece-by-sector provider PeerId from received bytes"
+                        );
+                    }
+                }
+                Ok(None) => {
+                    info!(%piece_index,?key, "get_value returned no piece-by-sector provider");
+                }
+                Err(err) => {
+                    error!(%piece_index,?key, ?err, "get_value returned an error (piece-by-sector)");
+                }
+            }
+        }
+
+        None
+    }
+}
+
+#[async_trait]
+impl<'a, RC: RpcClient> PieceReceiver for MultiChannelPieceReceiver<'a, RC> {
+    async fn get_piece(
+        &mut self,
+        piece_index: PieceIndex,
+    ) -> Result<Option<Piece>, Box<dyn Error + Send + Sync + 'static>>
+    where
+        RC: RpcClient,
+    {
+        trace!(%piece_index, "Piece request. DSN={:?}", self.dsn_node.is_some());
+
+        self.register_piece_index(piece_index);
+
+        if self.dsn_node.is_some() {
+            // until we get a valid piece
+            loop {
+                self.check_cancellation()?;
+
+                if let Some(piece) = self.get_piece_from_cache(piece_index).await {
+                    return Ok(Some(piece));
+                }
+
+                if let Some(piece) = self.get_piece_from_archival_storage(piece_index).await {
+                    return Ok(Some(piece));
+                }
+
+                sleep(Duration::from_secs(GET_PIECE_WAITING_DURATION_IN_SECS)).await;
```

Comment:
Let's at least print a warning so application doesn't appear to be stuck completely.